packages/node/src/lib/retry.spec.ts

```diff
@@ -0,0 +1,205 @@
+    nock(BACKEND_URL)
+      .post(TRIGGER_PATH)
+      .reply(500, { message: 'Server Exception' });
+
+    nock(BACKEND_URL)
+      .post(TRIGGER_PATH)
+      .times(10)
+      .reply(201, { acknowledged: true, transactionId: '1003' });
+
+    const idempotencyKeys: string[] = [];
+
+    for (let i = 0; i < 10; i++) {
+      const result = await novu.trigger('fake-workflow', {
+        to: { subscriberId: '123' },
+        payload: {},
+      });
+
+      idempotencyKeys.push(result.request?.headers['idempotency-key']);
+    }
+
+    expect(allEqual(idempotencyKeys)).toEqual(false);
+  });
+
+  it('should retry on status 422 and regenerate idempotency-key for every retry', async () => {
+    const idempotencyKeys: string[] = [];
+
+    nock(BACKEND_URL)
+      .post(TRIGGER_PATH)
+      .times(3)
+      .reply(function (_url, _body) {
+        idempotencyKeys.push(this.req.getHeader('idempotency-key') as string);
+
+        return [422, { message: 'Unprocessable Content' }];
+      });
+
+    nock(BACKEND_URL)
+      .post(TRIGGER_PATH)
+      .reply(201, { acknowledged: true, transactionId: '1003' });
+
+    const result = await novu.trigger('fake-workflow', {
+      to: { subscriberId: '123' },
+      payload: {},
+    });
+
+    // idempotency key should be regenerated for every retry for http status 422.
+    expect(allEqual(idempotencyKeys)).toBeFalsy();
+    expect(result.status).toEqual(201);
+    expect(result.request.headers['idempotency-key']).toBeDefined();
+  });
+
+  it('should retry getting topics list', async () => {
+    nock(BACKEND_URL)
+      .get(TOPICS_PATH)
+      .times(3)
+      .reply(500, { message: 'Server Exception' });
+
+    nock(BACKEND_URL).get(TOPICS_PATH).reply(200, [{}, {}]);
+
+    const result = await novu.topics.list({});
+
+    expect(result.status).toEqual(200);
+    expect(result.request.headers['idempotency-key']).toBeUndefined();
+  });
+
+  it('should fail after reaching max retries', async () => {
+    nock(BACKEND_URL)
+      .get(TOPICS_PATH)
+      .times(4)
+      .reply(500, { message: 'Server Exception' });
+
+    nock(BACKEND_URL).get(TOPICS_PATH).reply(200, [{}, {}]);
+
+    await expect(novu.topics.list({})).rejects.toMatchObject({
+      response: { status: 500 },
+    });
+  });
+
+  const NON_RECOVERABLE_ERRORS: Array<[number, string]> = [
+    [400, 'Bad Request'],
+    [401, 'Unauthorized'],
+    [403, 'Forbidden'],
+    [404, 'Not Found'],
+    [405, 'Method not allowed'],
+    [413, 'Payload Too Large'],
+    [414, 'URI Too Long'],
+    [415, 'Unsupported Media Type'],
+  ];
+
+  test.each<[number, string]>(NON_RECOVERABLE_ERRORS)(
+    'should not retry on non-recoverable %i error',
+    async (status, message) => {
+      nock(BACKEND_URL).get(TOPICS_PATH).times(3).reply(status, { message });
+      nock(BACKEND_URL).get(TOPICS_PATH).reply(200, [{}, {}]);
+
+      await expect(novu.topics.list({})).rejects.toMatchObject({
+        response: { status },
+      });
+    }
+  );
+
```

Comment:
Not sure if I were supposed to use jest-loop to test *non-recoverable* or *retryable* 4xx errors. In new changes I made it for non recoverable 4xx errors.